infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java

```diff
@@ -37,108 +38,66 @@
 @Slf4j
 public class SequencerServer extends AbstractServer {
 
-    /**
-     * A scheduler, which is used to schedule checkpoints and lease renewal
-     */
-    private final ScheduledExecutorService scheduler =
-            Executors.newScheduledThreadPool(
-                    1,
-                    new ThreadFactoryBuilder()
-                            .setDaemon(true)
-                            .setNameFormat("Seq-Checkpoint-%d")
-                            .build());
-    @Getter
-    long epoch;
-    AtomicLong globalLogTail;
+    private static final String PREFIX_SEQUENCER = "SEQUENCER";
+    private static final String KEY_SEQUENCER = "CURRENT";
 
     /**
-     * The file channel.
+     * Inherit from CorfuServer a server context
      */
-    private FileChannel fc;
-    private Object fcLock = new Object();
+    private final ServerContext serverContext;
 
     /**
      * Our options
      */
-    private Map<String, Object> opts;
-
-    /** Handler for this server */
-    @Getter
-    private CorfuMsgHandler handler = new CorfuMsgHandler()
-            .generateHandlers(MethodHandles.lookup(), this);
+    private final Map<String, Object> opts;
 
     /**
-     * map from stream-trails to global-log tails. used for backpointers.
+     * The sequencer maintains information about log and streams.
+     * Every append to the log updates the information in these maps.
+     *
+     * The following informaion and maps reflect the state of the logs and streams:
+     *
+     *  - globalLogTail: global log tail
+     *  - streamTailMap: a map of per-stream tail
+     *  - map from stream-tails to global-log tails. used for backpointers.
+     *  - cache of maxConflictCacheSize conflict keys and their latest commit (global-log) index
      */
-    ConcurrentHashMap<UUID, Long> streamTailToGlobalTailMap;
+    @Getter
+    private final AtomicLong globalLogTail = new AtomicLong(0L);
+    private final ConcurrentHashMap<UUID, Long> streamTailMap = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<UUID, Long> streamTailToGlobalTailMap = new ConcurrentHashMap<>();
+    private final long maxConflictCacheSize = 10000;
+    private final Cache<Object, Long> conflictToGlobalTailCache = Caffeine.newBuilder()
+            .maximumSize(maxConflictCacheSize)
+                .build();
 
     /**
-     * map of stream tails.
+     * A sequencer needs a lease to serve a certain number of tokens.
+     * The lease starting index is persisted.
+     * A lease is good for (@Link #SequencerServer::leaseLength) number of tokens.
+     *
+     * A lease is renewed when we reach leaseRenew tokens away from the limit.
      */
-    ConcurrentHashMap<UUID, Long> streamTailMap;
+    @Getter
+    private final long leaseLength = 10000;
```

Comment:
Can these settings come from ServerContext? You might have to modify the docopt.